File: app/src/main/java/net/mm2d/android/orientationfaker/MainActivity.java
/*
 * Copyright (c) 2014 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.android.orientationfaker;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v4.util.Pair;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.CheckBox;
import android.widget.TextView;

import net.mm2d.android.orientationfaker.orientation.OrientationHelper;
import net.mm2d.android.orientationfaker.orientation.OrientationIdManager;
import net.mm2d.android.orientationfaker.orientation.OrientationIdManager.OrientationId;
import net.mm2d.android.orientationfaker.orientation.OverlayPermissionHelper;
import net.mm2d.android.orientationfaker.settings.Settings;

import java.util.ArrayList;
import java.util.List;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class MainActivity extends AppCompatActivity {
    private static final String ACTION_UPDATE = "ACTION_UPDATE";
    private static final int REQUEST_CODE = 101;
    private Settings mSettings;
    private OrientationHelper mOrientationHelper;
    private TextView mStatusDescription;
    private CheckBox mResidentCheckBox;
    private List<Pair<Integer, View>> mButtonArray = new ArrayList<>();
    private Handler mHandler = new Handler(Looper.getMainLooper());

    private BroadcastReceiver mReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(final Context context, final Intent intent) {
            setStatusDescription();
            setOrientationIcon();
        }
    };

    public static void notifyUpdate(@NonNull final Context context) {
        LocalBroadcastManager.getInstance(context)
                .sendBroadcast(new Intent(ACTION_UPDATE));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mOrientationHelper = OrientationHelper.getInstance(this);
        mSettings = new Settings(this);
        mStatusDescription = findViewById(R.id.status_description);
        mResidentCheckBox = findViewById(R.id.resident_check_box);
        mResidentCheckBox.setClickable(false);

        findViewById(R.id.status).setOnClickListener(v -> toggleStatus());
        findViewById(R.id.resident).setOnClickListener(v -> toggleResident());
        findViewById(R.id.license).setOnClickListener(v -> {
            startActivity(new Intent(this, LicenseActivity.class));
        });
        setStatusDescription();
        setResidentCheckBox();
        setUpOrientationIcons();
        LocalBroadcastManager.getInstance(this)
                .registerReceiver(mReceiver, new IntentFilter(ACTION_UPDATE));

        if (mSettings.shouldResident()) {
            MainService.start(this);
        } else if (!OverlayPermissionHelper.canDrawOverlays(this)) {
            MainService.stop(this);
        }
        checkPermission();
    }

    private void checkPermission() {
        OverlayPermissionHelper.requestOverlayPermissionIfNeed(this, REQUEST_CODE);
    }

    @Override
    protected void onActivityResult(final int requestCode, final int resultCode, final Intent data) {
        if (requestCode == REQUEST_CODE) {
            mHandler.postDelayed(this::checkPermission, 1000);
            return;
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        LocalBroadcastManager.getInstance(this)
                .unregisterReceiver(mReceiver);
    }

    private void setUpOrientationIcons() {
        for (final OrientationId id : OrientationIdManager.getList()) {
            final View button = findViewById(id.getViewId());
            mButtonArray.add(new Pair<>(id.getOrientation(), button));
            button.setOnClickListener(v -> setOrientation(id.getOrientation()));
        }
        setOrientationIcon();
    }

    private void toggleStatus() {
        if (mOrientationHelper.isEnabled()) {
            MainService.stop(this);
            if (mSettings.shouldResident()) {
                mSettings.setResident(false);
                setResidentCheckBox();
            }
        } else {
            MainService.start(this);
        }
    }

    private void setStatusDescription() {
        final int stringRes = mOrientationHelper.isEnabled() ?
                R.string.status_running : R.string.status_waiting;
        mStatusDescription.setText(stringRes);
    }

    private void toggleResident() {
        mSettings.setResident(!mSettings.shouldResident());
        setResidentCheckBox();
        if (mSettings.shouldResident() && !mOrientationHelper.isEnabled()) {
            MainService.start(this);
        }
    }

    private void setResidentCheckBox() {
        mResidentCheckBox.setChecked(mSettings.shouldResident());
    }

    private void setOrientation(final int orientation) {
        mSettings.setOrientation(orientation);
        setOrientationIcon();
        if (mOrientationHelper.isEnabled()) {
            MainService.start(this);
        }
    }

    private void setOrientationIcon() {
        final Integer orientation = mSettings.getOrientation();
        for (final Pair<Integer, View> pair : mButtonArray) {
            if (orientation.equals(pair.first)) {
                pair.second.setBackgroundResource(R.drawable.bg_icon_selected);
            } else {
                pair.second.setBackgroundResource(R.drawable.bg_icon);
            }
        }
    }
}

File: app/src/main/java/net/mm2d/android/orientationfaker/MainActivity.kt
/*
 * Copyright (c) 2014 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.android.orientationfaker

import android.content.*
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.support.v4.content.LocalBroadcastManager
import android.support.v4.util.Pair
import android.support.v7.app.AppCompatActivity
import android.text.TextUtils
import android.text.format.DateFormat
import android.view.Menu
import android.view.MenuItem
import android.view.View
import kotlinx.android.synthetic.main.layout_main.*
import net.mm2d.android.orientationfaker.orientation.OrientationHelper
import net.mm2d.android.orientationfaker.orientation.OrientationIdManager
import net.mm2d.android.orientationfaker.orientation.OverlayPermissionHelper
import net.mm2d.android.orientationfaker.settings.Settings
import net.mm2d.log.Log
import java.util.*

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class MainActivity : AppCompatActivity() {
    private val settings by lazy {
        Settings.get()
    }
    private val orientationHelper by lazy {
        OrientationHelper.getInstance(this)
    }
    private val buttonList = ArrayList<Pair<Int, View>>()
    private val handler = Handler(Looper.getMainLooper())
    private val receiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            setStatusDescription()
            setOrientationIcon()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        supportActionBar?.title = getString(R.string.app_name)
        supportActionBar?.subtitle = makeVersionInfo()

        status.setOnClickListener { toggleStatus() }
        resident.setOnClickListener { toggleResident() }
        setStatusDescription()
        setResidentCheckBox()
        setUpOrientationIcons()
        registerReceiver()
        if (!OverlayPermissionHelper.canDrawOverlays(this)) {
            MainService.stop(this)
        } else if (settings.shouldResident()) {
            MainService.start(this)
        }
        checkPermission()
    }

    private fun checkPermission() {
        OverlayPermissionHelper.requestOverlayPermissionIfNeed(this, REQUEST_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_CODE) {
            handler.postDelayed({ checkPermission() }, 1000)
            return
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver()
    }

    private fun registerReceiver() {
        LocalBroadcastManager.getInstance(this)
                .registerReceiver(receiver, IntentFilter(ACTION_UPDATE))
    }

    private fun unregisterReceiver() {
        LocalBroadcastManager.getInstance(this)
                .unregisterReceiver(receiver)
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.license -> startActivity(Intent(this, LicenseActivity::class.java))
            R.id.play_store -> openGooglePlay(this)
        }
        return true
    }

    private fun setUpOrientationIcons() {
        OrientationIdManager.list.forEach {
            val orientation = it.orientation
            val button = findViewById<View>(it.viewId)
            buttonList.add(Pair(orientation, button))
            button.setOnClickListener { setOrientation(orientation) }
        }
        setOrientationIcon()
    }

    private fun toggleStatus() {
        if (orientationHelper.isEnabled) {
            MainService.stop(this)
            if (settings.shouldResident()) {
                settings.setResident(false)
                setResidentCheckBox()
            }
        } else {
            MainService.start(this)
        }
    }

    private fun setStatusDescription() {
        val enabled = orientationHelper.isEnabled
        statusSwitch.isChecked = enabled
        statusDescription.setText(if (enabled) R.string.status_running else R.string.status_waiting)
    }

    private fun toggleResident() {
        settings.setResident(!settings.shouldResident())
        setResidentCheckBox()
        if (settings.shouldResident() && !orientationHelper.isEnabled) {
            MainService.start(this)
        }
    }

    private fun setResidentCheckBox() {
        residentCheckBox.isChecked = settings.shouldResident()
    }

    private fun setOrientation(orientation: Int) {
        settings.orientation = orientation
        setOrientationIcon()
        if (orientationHelper.isEnabled) {
            MainService.start(this)
        }
    }

    private fun setOrientationIcon() {
        val orientation = settings.orientation
        for (pair in buttonList) {
            pair.second?.run {
                setBackgroundResource(if (orientation == pair.first) R.drawable.bg_icon_selected else R.drawable.bg_icon)
            }
        }
    }

    private fun makeVersionInfo(): String {
        return "Ver." + BuildConfig.VERSION_NAME +
                if (BuildConfig.DEBUG) " # " + DateFormat.format("yyyy/M/d kk:mm:ss", BuildConfig.BUILD_TIME)
                else ""
    }

    companion object {
        private const val PACKAGE_NAME = "net.mm2d.android.orientationfaker"
        private const val ACTION_UPDATE = "ACTION_UPDATE"
        private const val REQUEST_CODE = 101

        fun notifyUpdate(context: Context) {
            LocalBroadcastManager.getInstance(context)
                    .sendBroadcast(Intent(ACTION_UPDATE))
        }

        private fun openUri(context: Context, uri: String?): Boolean {
            if (TextUtils.isEmpty(uri)) {
                return false
            }
            try {
                val intent = Intent(Intent.ACTION_VIEW, Uri.parse(uri))
                intent.addCategory(Intent.CATEGORY_BROWSABLE)
                context.startActivity(intent)
            } catch (e: ActivityNotFoundException) {
                Log.w(e)
                return false
            }
            return true
        }

        private fun openGooglePlay(context: Context, packageName: String): Boolean {
            return openUri(context, "market://details?id=$packageName") ||
                    openUri(context, "https://play.google.com/store/apps/details?id=$packageName")
        }

        private fun openGooglePlay(context: Context): Boolean {
            return openGooglePlay(context, PACKAGE_NAME)
        }
    }
}
